src/backend/distributed/transaction/lock_graph.c

```diff
@@ -0,0 +1,772 @@
+ * concurrent lock acquisitions/releases.
+ */
+static void
+LockLockData(void)
+{
+	int partitionNum = 0;
+
+	LockDistributedTransactionSharedMemory(LW_SHARED);
+
+	for (partitionNum = 0; partitionNum < NUM_LOCK_PARTITIONS; partitionNum++)
+	{
+		LWLockAcquire(LockHashPartitionLockByIndex(partitionNum), LW_SHARED);
+	}
+}
+
+
+/*
+ * UnlockLockData unlocks the locks on the shared lock data structure in reverse
+ * order since LWLockRelease searches the given lock from the end of the
+ * held_lwlocks array.
+ */
+static void
+UnlockLockData(void)
+{
+	int partitionNum = 0;
+
+	for (partitionNum = NUM_LOCK_PARTITIONS - 1; partitionNum >= 0; partitionNum--)
+	{
+		LWLockRelease(LockHashPartitionLockByIndex(partitionNum));
+	}
+
+	UnlockDistributedTransactionSharedMemory();
+}
+
+
+/*
+ * AddEdgesForLockWaits adds an edge to the wait graph for every lock that
+ * waitingProc is waiting for.
+ */
+static void
+AddEdgesForLockWaits(WaitGraph *waitGraph, PGPROC *waitingProc, PROCStack *remaining)
+{
+	BackendData *waitingBackendData = GetBackendDataForProc(waitingProc);
+
+	/* the lock for which this process is waiting */
+	LOCK *waitLock = waitingProc->waitLock;
+
+	/* determine the conflict mask for the lock level used by the process */
+	LockMethod lockMethodTable = GetLocksMethodTable(waitLock);
+	int numLockModes = lockMethodTable->numLockModes;
+	int conflictMask = lockMethodTable->conflictTab[waitingProc->waitLockMode];
+
+	/* iterate through the queue of processes holding the lock */
+	SHM_QUEUE *procLocks = &(waitLock->procLocks);
+	PROCLOCK *procLock = (PROCLOCK *) SHMQueueNext(procLocks, procLocks,
+												   offsetof(PROCLOCK, lockLink));
+
+	while (procLock != NULL)
+	{
+		PGPROC *blockingProc = procLock->tag.myProc;
+
+		/* a proc never blocks itself or any other lock group member */
+		if (!IsSameLockGroup(waitingProc, blockingProc))
+		{
+			BackendData *blockingBackendData = GetBackendDataForProc(blockingProc);
+
+			int lockMode = 1;
+
+			/* check conflicts with every locktype held */
+			for (; lockMode <= numLockModes; lockMode++)
+			{
+				if ((procLock->holdMask & LOCKBIT_ON(lockMode)) &&
+					(conflictMask & LOCKBIT_ON(lockMode)))
+				{
+					bool isBlockingXactWaiting = IsProcessWaitingForLock(blockingProc);
+					if (isBlockingXactWaiting)
+					{
+						remaining->procs[remaining->procCount++] = blockingProc;
+					}
+
+					AddWaitEdge(waitGraph, waitingBackendData, waitingProc,
+								blockingBackendData, blockingProc, isBlockingXactWaiting);
+				}
+			}
+		}
+
+		procLock = (PROCLOCK *) SHMQueueNext(procLocks, &procLock->lockLink,
+											 offsetof(PROCLOCK, lockLink));
+	}
+}
+
+
+/*
+ * AddEdgesForWaitQueue adds an edge to the wait graph for processes in front
+ * of waitingProc in the wait queue that are trying to acquire a conflicting
+ * lock.
+ */
+static void
+AddEdgesForWaitQueue(WaitGraph *waitGraph, PGPROC *waitingProc, PROCStack *remaining)
+{
```

Comment:
Waiting PROCLOCKs have 0 as their holdMask (since they don't hold anything), which means that we can't check for conflict.

There's probably a way to simplify postgres though :)